### DISCUSSION TITLE
Is it possible to view raw glyph data?

### DISCUSSION THREAD

**Discussion by hanakp-VUT** [fontforge/fontforge#5424] [Q&A] [answered]
Hello, I wish to view raw glyph data. Specifically, I'm analyzing Type1 fonts embedded in PDF files and I want to see what is actually stored for each glyph in PDFStream. Is it possible to do that in FontForge (or other program)? My google searches came up empty. Thanks.

**Comment by 2-www** [comment]
i'm not sure if that's what you need, but with [fonttools' ttx](https://fonttools.readthedocs.io/en/latest/ttx.html) you can convert a font file to an xml file with entries like this:
```
          <CharString index="46">
            -143 65 -103 122 vhcurveto
            -107 253 rmoveto
            return
          </CharString>
```

**Reply by hanakp-VUT** [reply]
Thanks for the tip. It works, but there is one complication - TTX can't open Postscript Type1 fonts, it throws "unknown file type" error. I have to use Fontforge to convert them to OTF first and then use TTX to convert them to XML. I'm a bit worried that such double conversion may introduce rounding errors or other imprecisions. My aim was to compare glyph shapes of identical fonts between similar documents, i.e. if they were affected by hinting or other typographical tricks. But now I can't be sure what introduced the differences - the original authoring program or the conversions.

I know there are libraries (such as FreeType) that can parse Postscipt fonts (their /CharStrings dictionary) directly, but I'd like to avoid writing a program for one-off analysis. How Fontforge draws the glyphs in the glyph shape editor? Is there a way to export the drawing coordinates and/or instructions somehow?

**Reply by skef** [reply, marked as answer]
Adobe's [AFDKO](https://github.com/adobe-type-tools/afdko) `tx` program can dump Type 1 glyph information directly. E.g. `tx -dump -6 [filename]` . Use `-g` to pick particular glyphs to view.

**Reply by hanakp-VUT** [reply]
Thanks, it works great! Although tx actually does one internal conversion too, as explained in its help:
_"Font data is read from input, converted to ABF, which is then converted for output."_

However, I encountered another problem: Fontforge refuses to load about half of PDF files I tried, it throws errors
_"xxx.pdf is not in a known format (or uses features of that format fontforge does not support, or is so badly corrupted as to be unreadable"_. In case anyone else runs into it, here is what I tried: 

- Mutool with extract option works, but it names the fonts by their PDF object number, which is not very practical. Fortunately, it's easy to search the extracted files for the desired font's name. 
- Trial version of VeryPDF pdffont utility produces separate AFM and CFF files that can't be read by tx ("_can't read source stream_" error).
- This [online extractor](https://www.konwerter.net/en/ExtractPdfAttachments/) worked best for me so far, it produces "correct" CFF files with fonts' names. Obviously, it can't be used for sensitive documents.
- More options are listed in [this Substack thread](https://stackoverflow.com/questions/3488042/how-can-i-extract-embedded-fonts-from-a-pdf-as-valid-font-files), although they either do extra font conversions or are too complicated.
